lib/spack/spack/analyzers/base.py

```diff
@@ -0,0 +1,70 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+"""An analyzer base provides basic functions to run the analysis, save results,
+and (optionally) interact with a Spack Monitor
+"""
+
+import spack.monitor
+import llnl.util.tty as tty
+
+import os
+
+
+class AnalyzerBase:
+
+    def __init__(self, spec, dirname=None):
+        """An Analyzer is intended to run on one spec install, so the spec
+        with its associated package is required on init. The child analyzer
+        class should define an init function that super's the init here, and
+        also check that the analyzer has all dependencies that it
+        needs. If an analyzer subclass does not have dependencies, it does not
+        need to define an init. An Analyzer should not be allowed to proceed
+        if one or more dependencies are missing. The dirname, if defined,
+        is an optional directory name to save to (instead of the default meta
+        spack directory).
+        """
+        self.spec = spec
+        self.dirname = dirname
+
+        # An analyzer cannot be run if the spec isn't associated with a package
+        if not hasattr(spec, "package") or not spec.package:
+            tty.die("A spec can only be analyzed with an associated package.")
+        self.meta_dir = os.path.dirname(spec.package.install_log_path)
+
+        for required in ["name", "outfile", "description"]:
+            if not hasattr(self, required):
+                tty.die("Please add a %s attribute on the analyzer." % required)
+
+    def run(self):
+        """Given a spec with an installed package, run the analyzer on it.
+        """
+        raise NotImplementedError
+
+    def get_outfile(self, dirname=None, outfile=None):
```

Comment:
can this be an `outfile` property instead of `get_outfile`?

Reply:
The reason I have it as a function instead of a property (and have it named `get_outfile`) is because it accepts arguments that would determine logic for getting the output file (e.g., the directory or a different output file name). I think a property would be more appropriate if this is already set. So either we leave as a function, or we require that the user provide the exact outdir / dirname on init of the class, and then use those. Let me know your preference!

Reply:
yeah I guess this is just in the abigail analyzer, so let's keep it.  Shoudl it be a private method though?

Reply:
Good thinking, I'll make it private.

Reply:
Okay, so I also wasn't happy with this still - since we have consistency with an output directory (but not the file) I created a property for the output_dir, which gets generated once and then "cached" at `self._output_dir`. Then the developer user can just `os.path.join` that directory with whatever output file basename they want. I also made it so each analyzer has it's own folder so we don't have to worry about name conflicts.